### ISSUE TITLE
Create semantic conventions for API Gateway

### ISSUE THREAD

**Issue by SonjaChevre** [open-telemetry/semantic-conventions#183]
We (Tyk Technologies, maintainer of the [Tyk open source API Gateway](https://github.com/TykTechnologies/tyk)) would like to work on adding a semantic conventions for API Gateways.

### What is an API Gateway?

An API gateway is a tool that aggregates unique application APIs, making them all available in one place. It allows organizations to move key functions, such as authentication and authorization or limiting the number of requests between applications, to a centrally managed location. An API gateway functions as a common interface to (often external) API consumers. From: [API Gateway](https://glossary.cncf.io/api-gateway/) 
 

### What are specific observability needs with APIs and API Gateways?

**1. Need to generate RED metrics per API name and API version**

API teams need to be able to track request rates, error rates and duration per API and per API version. 

**2. Need to configure different observability pipeline depending on the API name or API tag**

Because an API Gateway is a central component that processes traffic for many teams, it is often the case that it captures data that are relevant to different teams, each having their own observability need (different observability back-end, different need for sampling, different need to remove information, …).

### What is the current support of API Gateways in OTel?

There are to our knowledge no semantic conventions specific to APIs or API Gateways at the moment.

### What are we missing in the semantic conventions?

Non exhaustive list:

- API name
- API version

### What is the suggested approach?

We are actively working on adding this information in our Tyk API Gateway ([GitHub - TykTechnologies/tyk: Tyk Open Source API Gateway written in Go, supporting REST, GraphQL, TCP and gRPC protocols](https://github.com/TykTechnologies/tyk) )   and would welcome other members of the observability and API communities to join us on improving the semantic conventions.

Looking forward to see if this proposal gets any interest!

Sonja

Note: we are also working on another proposal to introduce semantic conventions for GraphQL: https://github.com/open-telemetry/semantic-conventions/issues/182

**Comment by joaopgrassi** [comment]
This was closed by [mistake in the stale bot.](https://github.com/open-telemetry/semantic-conventions/pull/771) Re-opening

**Comment by drazzib** [comment]
As described in initial issue, OpenTelemetry SemConv lacks specific attributes to describe the business and management context of API interactions managed by a API Gateway (like GCP Apigee, Amazon Gravitee Gateway...)

I would like to propose adding the following attributes.

> These would typically be applied to **spans representing requests handled** by an API Gateway.
> For example, a user could easily build a dashboard showing the error rate for all APIs associated with a specific subscription plan, regardless of the underlying gateway technology.

### `api.id`

- Type: UUID
- Description: A unique identifier for the API being invoked. This is the identifier of the API definition or product as configured in the API Gateway. It should be a stable, human-readable identifier.
- Example: `1125f9c3-59cd-49ab-ba89-731f97666b59`

### `api.plan.name`

- Type: string
- Description: The identifier for the API consumption plan or tier associated with the request. This is useful for tracking usage against different service levels (e.g., "Free", "Premium").
- Example: `premium-tier`, `free-plan-2025`

### `api.subscription.id`

- Type: string
- Description: A unique identifier for the specific subscription linking an application to an API product or plan.
- Example: `sub_1a2b3c4d5e6f`

### `api.application.id`

- Type: string
- Description: A unique identifier for the client application making the request, as registered in the API Gateway. This is often the "client ID" or a similar application-level identifier.
- Example: `app_f6e5d4c3b2a1`

Is this something that might be of interest to the OpenTelemetry SemConv community?

**Comment by thompson-tomo** [comment]
I agree it would be beneficial to have defined conventions for api gateways however the above attributes doesn't seem to me to be the best solution.

What I am thinking is that we introduce:

- `destination.tag` to describe the machine the request is going to or `destination.app.name` for the application the request is going to. This is instead of `api.id`. My Preference would be to actually offer both.
- `source.tag` to describe the machine the request has come from or `source.app.name` to describe the application making the request. This is instead of `api.application.id`. My Preference would be the tag option as the app name would be too hard to extract without straying into auth.

I would leave out the other 2 attributes proposed as they seem to be more about auth/identity.

The key thing though would be to define an internal span for processing the request and clearly defining the links to the network level calls defined via #2873 in the documentation.

In the end I could see an api gateway request over http generating 4 spans:

- http server (network layer): captures entire operation and is the root span.
- api gateway internal (logical layer): captures processing done by gateway prior to forwarding request
- http client (network layer): captures call to the service
- api gateway internal (logical layer): captures processing done by gateway after getting response to forwarding request
